File: 4_textbook_track/1E_MinimumSkew.R
#!/usr/bin/env Rscript

getSkew <- function(fasta) {
    
    ## ----- Get header ----- ##
    header <- substring(fasta[1], 5, 12)
    
    ## ----- Process File ----- ##
    genome <- paste(fasta[2:length(fasta)], collapse="")
    
    ## Get genome length
    len <- nchar(genome)
    
    ## Split genome into characters
    nucleo = substring(fasta, 1:len, 1:len)
    
    ## ------ Get Results ------ ##
    
    ## Create data frame to hold results
    results <- vector(length=len)
    skew <- 0
    
    ## Loop through each nucleotide
    for(i in 1:len) {
        if(nucleo[i] == "C") {
              skew <- skew-1
        }
        else if(nucleo[i] == "G") {
              skew <- skew+1
        }
        
        results[i] <- skew
    }
    
    write.table(results, file=paste(header,"_skew.txt", sep=""))
}

minSkew <- function(x) {
    skew <- read.table(x)
    
    answer <- paste("Min:", min(skew), "at position", which.min(skew), ",", 
                    "Max:", max(skew), "at position", which.max(skew))
}

File: 4_textbook_track/1E_MinimumSkew.py
#!/usr/bin/env python3

# Find a position in a genome minimizing the skew.
# Given: A DNA string Genome.
# Return: All integer(s) i minimizing Skew(Prefix,i (Text)) over all values of i (from 0 to |Genome|).

'''
example:
   C  A  T G G G C A T C G G C C A T A  C G  C  C
0 -1 -1 -1 0 1 2 1 1 1 0 1 2 1 0 0 0 0 -1 0 -1 -2
'''

def find_min_skew(genome):
    counts = {}
    skew = 0
    
    for i in range(len(genome)):
        if genome[i] == 'C':
            skew -= 1
        elif genome[i] == 'G':
            skew += 1

        counts[i+1] = skew
            
    return counts

# Read file
with open('rosalind_1e.txt', 'r') as in_file:
        genome = in_file.read().rstrip()

# Find the smallest skew value and associated value(s) of i
skews = find_min_skew(genome)
min_skew = min(skews.values())
answer = [key for key, val in skews.items() if val == min_skew]
for a in answer: print a
